### PR TITLE
Drop connection from peers if invalid data was sent

### PR DESCRIPTION
**Pull request by iulianpascalau**
- a peer that sends invalid message (that can not be unmarshaled) will be directly added to the black list
- implemented unjoin functionality (pending fixing tests)

### CODE REVIEW

File: consensus/spos/worker.go [hunk 1]
```diff
@@ -325,6 +325,13 @@ func (wrk *Worker) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedP
 	cnsMsg := &consensus.Message{}
 	err = wrk.marshalizer.Unmarshal(cnsMsg, message.Data())
 	if err != nil {
+		//this situation is so severe that we have to black list both the message originator and the connected peer
+		//that disseminated this message.
+
+		reason := "blacklisted due to invalid consensus message"
```
Comment (bogdan-rosianu) [inline consensus/spos/worker.go:331]: maybe define a constant somewhere

File: consensus/spos/worker.go [hunk 2]
```diff
@@ -325,6 +325,13 @@ func (wrk *Worker) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedP
...
+		wrk.antifloodHandler.BlacklistPeer(message.Peer(), reason, core.IntervalBlackListPeerInvalidMessage)
+		wrk.antifloodHandler.BlacklistPeer(fromConnectedPeer, reason, core.IntervalBlackListPeerInvalidMessage)
```
Comment (bogdan-rosianu) [inline consensus/spos/worker.go:333]: as discussed, please make sure a peer that sends you the message but it's not the message's originator won't get blacklisted if it did nothing wrong.

File: core/constants.go [hunk 3]
```diff
@@ -445,3 +445,7 @@ const CommitMaxTime = time.Second
 
 // DefaultUnstakedEpoch represents the default epoch that is set for a validator that has not unstaked yet
 const DefaultUnstakedEpoch = math.MaxUint32
+
+// IntervalBlackListPeerInvalidMessage time to keep a peer in the black list if it sends a message that does not follow
```
Comment (bogdan-rosianu) [inline core/constants.go:449]: `represents the time...`

Comment (bogdan-rosianu) [reply core/constants.go:449]: also maybe rename to `InvalidMessageBlacklistDuration` or something similar. It's not an interval.

File: dataRetriever/resolvers/messageProcessor.go [hunk 4]
```diff
@@ -38,10 +38,15 @@ func (mp *messageProcessor) canProcessMessage(message p2p.MessageP2P, fromConnec
 }
 
 // parseReceivedMessage will transform the received p2p.Message in a RequestData object.
-func (mp *messageProcessor) parseReceivedMessage(message p2p.MessageP2P) (*dataRetriever.RequestData, error) {
+func (mp *messageProcessor) parseReceivedMessage(message p2p.MessageP2P, fromConnectedPeer core.PeerID) (*dataRetriever.RequestData, error) {
 	rd := &dataRetriever.RequestData{}
 	err := rd.UnmarshalWith(mp.marshalizer, message)
 	if err != nil {
+		//this situation is so severe that we need to black list de peers
```
Comment (bogdan-rosianu) [inline dataRetriever/resolvers/messageProcessor.go:45]: the *

File: process/interceptors/singleDataInterceptor.go [hunk 5]
```diff
@@ -77,6 +77,12 @@ func (sdi *SingleDataInterceptor) ProcessReceivedMessage(message p2p.MessageP2P,
 	interceptedData, err := sdi.factory.Create(message.Data())
 	if err != nil {
 		sdi.throttler.EndProcessing()
+
+		//this situation is so severe that we need to black list de peers
```
Comment (bogdan-rosianu) [inline process/interceptors/singleDataInterceptor.go:81]: the *

File: process/throttle/antiflood/p2pAntiflood.go [hunk 6]
```diff
@@ -183,6 +184,26 @@ func (af *p2pAntiflood) SetDebugger(debugger process.AntifloodDebugger) error {
 	return nil
 }
 
+// BlacklistPeer will add a peer to the black list
+func (af *p2pAntiflood) BlacklistPeer(peer core.PeerID, reason string, duration time.Duration) {
```
Comment (bogdan-rosianu) [inline process/throttle/antiflood/p2pAntiflood.go:188]: in the future if we found enough reasons to make this worth, we can define a set of reasons and the blacklist duration associated.

File: consensus/spos/worker.go (hunk 2 above)
Comment (iulianpascalau) [reply consensus/spos/worker.go:333]: Good catch, implemented unjoin, pending the integration test fixing

File: process/throttle/antiflood/p2pAntiflood.go (hunk 6 above)
Comment (iulianpascalau) [reply process/throttle/antiflood/p2pAntiflood.go:188]: Agree, first let's see what we need

File: consensus/spos/worker.go (hunk 1 above)
Comment (iulianpascalau) [reply consensus/spos/worker.go:331]: used only here, once, will do if I do need it elsewhere

File: consensus/spos/worker.go (hunk 2 above)
Comment (iulianpascalau) [reply consensus/spos/worker.go:333]: Done testing, removed one antiflood bug because of start in epoch mechanism

File: core/constants.go (hunk 3 above)
Comment (iulianpascalau) [reply core/constants.go:449]: done, renamed to InvalidMessageBlacklistDuration

File: dataRetriever/resolvers/messageProcessor.go (hunk 4 above)
Comment (iulianpascalau) [reply dataRetriever/resolvers/messageProcessor.go:45]: de :)))

File: process/interceptors/singleDataInterceptor.go (hunk 5 above)
Comment (iulianpascalau) [reply process/interceptors/singleDataInterceptor.go:81]: changed

File: p2p/issues/pubsub_349/main.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
+package main
```
Comment (bogdan-rosianu) [inline p2p/issues/pubsub_349/main.go:1]: do we need this committed?

File: p2p/libp2p/netMessenger.go [hunk 8]
```diff
@@ -729,11 +736,34 @@ func (netMes *networkMessenger) UnregisterAllMessageProcessors() error {
...
+	var errFound error
+	for topicName, t := range netMes.topics {
+		subscr := netMes.subscriptions[topicName]
+		if subscr != nil {
+			subscr.Cancel()
+		}
+
+		err := t.Close()
+		if err != nil {
+			errFound = err
```
Comment (bogdan-rosianu) [inline p2p/libp2p/netMessenger.go:758]: please log the errors. The way it is now we will only know the last error

File: p2p/issues/pubsub_349/main.go (hunk 7 above)
Comment (iulianpascalau) [reply p2p/issues/pubsub_349/main.go:1]: yes, it is a proof of the issue and a boiler plate for future issues (not to rewrite most of the code over and over and over again)

File: p2p/libp2p/netMessenger.go (hunk 8 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go:762]: done

**Pull request merged**